# First Sprint

## Data Cleaning

### Import necessary libraries

In [76]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from ast import literal_eval

### Read the data

#### Credits

In [18]:
credits = pd.read_csv('credits.csv')

In [19]:
credits.sample(5)

cast  \
31344  [{'cast_id': 5, 'character': 'Piera', 'credit_...   
19620  [{'cast_id': 3, 'character': 'Tyrone', 'credit...   
41123  [{'cast_id': 1001, 'character': 'Boyfriend/Wri...   
23516  [{'cast_id': 1, 'character': 'Axel', 'credit_i...   
298    [{'cast_id': 1, 'character': 'Margot', 'credit...   

                                                    crew      id  
31344  [{'credit_id': '52fe4882c3a36847f816aeeb', 'de...   53914  
19620  [{'credit_id': '52fe4a88c3a368484e158d63', 'de...  122928  
41123  [{'credit_id': '56e6fe8c92514122ba000fb3', 'de...   79871  
23516  [{'credit_id': '53a56270c3a3687a40000e4e', 'de...  220820  
298    [{'credit_id': '52fe43749251416c75010fb5', 'de...   10452

In [99]:
print(credits.shape[0], credits.shape[1])

45476 3


Dataframe "credits" has 45,476 rows and 3 columns

#### Keywords

In [20]:
keywords = pd.read_csv('keywords.csv')

In [21]:
keywords.sample(5)

id                                           keywords
36265  262199          [{'id': 9716, 'name': 'stand-up comedy'}]
32529  354556  [{'id': 2106, 'name': 'cold war'}, {'id': 9715...
34060   92819  [{'id': 156084, 'name': 'bricklayer'}, {'id': ...
44208  337339  [{'id': 339, 'name': 'submarine'}, {'id': 9663...
40630  397365                                                 []

In [103]:
print(keywords.shape[0], keywords.shape[1])

46419 2


Dataframe "keywords" has 46,419 rows and 2 columns

#### Links

In [22]:
links = pd.read_csv('links.csv')

In [23]:
links.sample(5)

movieId   imdbId    tmdbId
9758     31680   361322  193704.0
18140    90545    51114   83354.0
23974   112655  1763264  123109.0
14270    71376   479354   25532.0
43955   171639   249934   69072.0

In [104]:
print(links.shape[0], links.shape[1])

45843 3


Dataframe "links" has 45,843 rows and 3 columns

#### Links_small

In [25]:
links_small = pd.read_csv('links_small.csv')

In [26]:
links_small.sample(5)

movieId   imdbId    tmdbId
8967   140152  3215846  318033.0
8269   100383  2053463  109421.0
5039     7173   343135    5966.0
8044    94325  1327194   77877.0
2953     3699    88172    9663.0

In [105]:
print(links_small.shape[0], links_small.shape[1])

9125 3


Dataframe "links_small" has 9,125 rows and 3 columns

#### Movies

In [27]:
movies = pd.read_csv('movies_metadata.csv')

C:\Users\Dana\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


While reading our movies_metadata dataset, we received a warning "Columns (10) have mixed types". We are offered to specify dtype or to set low_memory to false. The last option seems to be the easiest one to do.

In [112]:
movies = pd.read_csv('movies_metadata.csv', parse_dates = ['release_date'], low_memory = False)

In [113]:
movies.sample(5)

adult belongs_to_collection  budget  \
26206  False                   NaN       0   
14960  False                   NaN       0   
25610  False                   NaN       0   
7639   False                   NaN  350000   
24250  False                   NaN       0   

                                                  genres  \
26206                                                 []   
14960  [{'id': 80, 'name': 'Crime'}, {'id': 18, 'name...   
25610  [{'id': 99, 'name': 'Documentary'}, {'id': 104...   
7639   [{'id': 27, 'name': 'Horror'}, {'id': 878, 'na...   
24250  [{'id': 18, 'name': 'Drama'}, {'id': 99, 'name...   

                               homepage      id    imdb_id original_language  \
26206  http://www.hbo.com/sports/mantle  142897  tt0470857                en   
14960                               NaN   80595  tt0054898                en   
25610                               NaN   82676  tt0840358                en   
7639                                NaN   11417  tt0087622                nl   
24250                               NaN  250666  tt3263996                en   

                                  original_title  \
26206                                     Mantle   
14960                        The Frightened City   
25610  Everyday Sunshine:  The Story of Fishbone   
7639                                     De Lift   
24250                           The Overnighters   

                                                overview  ... release_date  \
26206  The almost mythic career of Yankee slugger Mic...  ...   2005-07-13   
14960  A London accountant (Herbert Lom) forms six ga...  ...   1961-09-01   
25610  .From the shifting faultlines of Hollywood fan...  ...   2011-10-07   
7639   A lift begins displaying some erratic behavior...  ...   1983-05-11   
24250  Desperate, broken men chase their dreams and r...  ...   2014-10-10   

      revenue runtime                             spoken_languages    status  \
26206     0.0    60.0                                           []  Released   
14960     0.0    97.0     [{'iso_639_1': 'en', 'name': 'English'}]  Released   
25610     0.0   107.0     [{'iso_639_1': 'en', 'name': 'English'}]  Released   
7639      0.0    95.0  [{'iso_639_1': 'nl', 'name': 'Nederlands'}]  Released   
24250     0.0   102.0     [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                                 tagline  \
26206              The Definitive Story of Mickey Mantle   
14960                 A city terrorised by its own evil!   
25610                                                NaN   
7639   Take the stairs, take the stairs. For God's sa...   
24250                                                NaN   

                                           title  video vote_average  \
26206                                     Mantle  False          0.0   
14960                        The Frightened City  False          6.5   
25610  Everyday Sunshine:  The Story of Fishbone  False          8.0   
7639                                    The Lift  False          5.6   
24250                           The Overnighters  False          7.3   

      vote_count  
26206        0.0  
14960        2.0  
25610        1.0  
7639        26.0  
24250       29.0  

[5 rows x 24 columns]

In [106]:
print(movies.shape[0], movies.shape[1])

45466 24


Dataframe "movies" has 45,466 rows and 24 columns

#### Ratings

In [30]:
ratings = pd.read_csv('ratings.csv')

In [32]:
ratings.sample(5)

userId  movieId  rating   timestamp
4301196    44604     7026     2.5  1107544567
1703366    17741     1682     4.0  1162189791
19856445  206373     2700     3.0  1234056056
22004058  228563     7162     4.0  1093042865
10612859  109527     2011     3.0  1095879606

In [107]:
print(ratings.shape[0], ratings.shape[1])

26024289 4


Dataframe "ratings" has 26,024,289 rows and 4 columns

#### Ratings_small

In [33]:
ratings_small = pd.read_csv('ratings_small.csv')

In [35]:
ratings_small.sample(5)

userId  movieId  rating   timestamp
74185     518      546     1.0   945370191
63257     457     2762     3.0  1471384759
50514     371   115170     4.0  1462868738
31865     232     1235     4.0   955090035
8982       57     2245     4.0   907764671

In [108]:
print(ratings_small.shape[0], ratings_small.shape[1])

100004 4


Dataframe "ratings_small" has 100,004 rows and 4 columns

 ### Identify the type of variables in datasets

#### Credits

In [60]:
credits.dtypes

cast    object
crew    object
id       int64
dtype: object

#### Keywords

In [93]:
keywords.dtypes

id           int64
keywords    object
dtype: object

#### Links

In [95]:
links.dtypes

movieId      int64
imdbId       int64
tmdbId     float64
dtype: object

#### Links_small

In [96]:
links_small.dtypes

movieId      int64
imdbId       int64
tmdbId     float64
dtype: object

Change "tmdbId" column data type to integer

#### Movies

In [115]:
movies.dtypes

adult                     object
belongs_to_collection     object
budget                    object
genres                    object
homepage                  object
id                        object
imdb_id                   object
original_language         object
original_title            object
overview                  object
popularity                object
poster_path               object
production_companies      object
production_countries      object
release_date              object
revenue                  float64
runtime                  float64
spoken_languages          object
status                    object
tagline                   object
title                     object
video                     object
vote_average             float64
vote_count               float64
dtype: object

Change: 
- "budget" from object to float64
- "id", "imdb_id"

In [110]:
movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2      A family wedding reignites the ancient feud be...  ...   1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4      Just when George Banks has recovered from his ...  ...   1995-02-10   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisation of homosexual...  ...   2017-06-09   

           revenue runtime                                   spoken_languages  \
0      373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1      262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2              0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}

#### Ratings

In [97]:
ratings.dtypes

userId         int64
movieId        int64
rating       float64
timestamp      int64
dtype: object

#### Ratings_small

In [98]:
ratings_small.dtypes

userId         int64
movieId        int64
rating       float64
timestamp      int64
dtype: object

## Data Analysis

### Check for null values

#### Credits

In [86]:
credits.isna().sum()

cast    0
crew    0
id      0
dtype: int64

#### Keywords

In [87]:
keywords.isna().sum()

id          0
keywords    0
dtype: int64

#### Links

In [88]:
links.isna().sum()

movieId      0
imdbId       0
tmdbId     219
dtype: int64

#### Links_small

In [89]:
links_small.isna().sum()

movieId     0
imdbId      0
tmdbId     13
dtype: int64

#### Movies

In [90]:
movies.isna().sum()

adult                        0
belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
tagline                  25054
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

#### Ratings

In [91]:
ratings.isna().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

#### Ratings_small

In [92]:
ratings_small.isna().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

In [118]:
def get_names(x, l = 5):
    if isinstance(x, list):
        names = [i['name'] for i in x]
        if len(names) > l:
            names = names[ : l]
        return names
    return []

In [117]:
movies['genres'] = movies['genres'].apply(literal_eval)
movies['genres'] = movies['genres'].apply(get_names)

In [81]:
movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2      A family wedding reignites the ancient feud be...  ...   1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4      Just when George Banks has recovered from his ...  ...   1995-02-10   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisation of homosexual...  ...   2017-06-09   

           revenue runtime                                   spoken_languages  \
0      373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1      262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2              0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}

In [82]:
movies['belongs_to_collection'] = movies['belongs_to_collection'].apply(literal_eval)
movies['belongs_to_collection'] = movies['belongs_to_collection'].apply(get_names)

ValueError: malformed node or string: nan